agent/agentcontainers/containers_dockercli.go

```diff
@@ -31,6 +33,172 @@ func NewDocker(execer agentexec.Execer) Lister {
+func EnvInfo(ctx context.Context, execer agentexec.Execer, container, containerUser string) (*ContainerEnvInfoer, error) {
+	var dei ContainerEnvInfoer
+	dei.container = container
+
+	var stdoutBuf, stderrBuf bytes.Buffer
+	if containerUser == "" {
+		// Get the "default" user of the container if no user is specified.
+		// TODO: handle different container runtimes.
+		cmd, args := WrapDockerExec(container, "")("whoami")
+		execCmd := execer.CommandContext(ctx, cmd, args...)
+		execCmd.Stdout = &stdoutBuf
+		execCmd.Stderr = &stderrBuf
+		if err := execCmd.Run(); err != nil {
+			return nil, xerrors.Errorf("get container user: run whoami: %w: stderr: %q", err, strings.TrimSpace(stderrBuf.String()))
+		}
+		out := strings.TrimSpace(stdoutBuf.String())
+		if len(out) == 0 {
+			return nil, xerrors.Errorf("get container user: run whoami: empty output: stderr: %q", strings.TrimSpace(stderrBuf.String()))
+		}
+		containerUser = out
+		stdoutBuf.Reset()
+		stderrBuf.Reset()
+	}
+	// Now that we know the username, get the required info from the container.
+	// We can't assume the presence of `getent` so we'll just have to sniff /etc/passwd.
+	cmd, args := WrapDockerExec(container, containerUser)("cat", "/etc/passwd")
+	execCmd := execer.CommandContext(ctx, cmd, args...)
+	execCmd.Stdout = &stdoutBuf
+	execCmd.Stderr = &stderrBuf
+	if err := execCmd.Run(); err != nil {
+		return nil, xerrors.Errorf("get container user: read /etc/passwd: %w stderr: %q", err, strings.TrimSpace(stderrBuf.String()))
+	}
+
+	scanner := bufio.NewScanner(&stdoutBuf)
+	var foundLine string
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if len(line) == 0 {
+			continue
+		}
+		if !strings.HasPrefix(line, containerUser+":") {
+			continue
+		}
+		foundLine = line
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, xerrors.Errorf("get container user: scan /etc/passwd: %w", err)
+	}
+	if foundLine == "" {
+		return nil, xerrors.Errorf("get container user: no matching entry for %q found in /etc/passwd", containerUser)
+	}
+
+	// Parse the output of /etc/passwd. It looks like this:
+	// postgres:x:999:999::/var/lib/postgresql:/bin/bash
+	passwdFields := strings.Split(foundLine, ":")
+	if len(passwdFields) < 7 {
+		return nil, xerrors.Errorf("get container user: invalid line in /etc/passwd: %q", foundLine)
+	}
+
+	// The fourth entry in /etc/passwd contains GECOS information, which is a
+	// comma-separated list of fields. The first field is the user's full name.
+	gecos := strings.Split(passwdFields[4], ",")
+	fullName := ""
+	if len(gecos) > 1 {
+		fullName = gecos[0]
+	}
+
+	dei.user = &user.User{
+		Gid:      passwdFields[3],
+		HomeDir:  passwdFields[5],
+		Name:     fullName,
+		Uid:      passwdFields[2],
+		Username: containerUser,
+	}
+	dei.userShell = passwdFields[6]
+
+	// Finally, get the environment of the container.
+	stdoutBuf.Reset()
+	stderrBuf.Reset()
+	cmd, args = WrapDockerExec(container, containerUser)("env")
```

Comment:
If we're only using env here, do we actually need to fetch the container environment for anything? I assume it'll be automatically applied to any command that is run anyway.

I also believe this method may also be missing environment variables that should be set according to the devcontainer.json.

Reply:
You mean `remoteEnv`, correct?

Reply:
Ref: https://code.visualstudio.com/docs/devcontainers/attach-container

It looks like it can be one of `none`, `interactiveShell`, `loginShell`, or `loginInteractiveShell`. I think I may add a TODO for now and address this in a follow-up PR.

Reply:
OK, I ended up implementing this. It's already in its own package but there's probably some more refactoring to be done here.